src/server/pegasus_hotkey_collector.cpp

```diff
@@ -0,0 +1,175 @@
+{
+    if (_collector_state.load() == STOP || count == 0) {
+        return;
+    }
+    for (int i = 0; i < count; i++) {
+        ::dsn::blob key;
+        if (requests[i] != nullptr) {
+            dsn::task_code rpc_code(requests[0]->rpc_code());
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_MULTI_PUT) {
+                dsn::apps::multi_put_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.hash_key;
+                capture_str_data(key.to_string());
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_INCR) {
+                dsn::apps::incr_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.key;
+                capture_blob_data(key);
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_CHECK_AND_SET) {
+                dsn::apps::check_and_set_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.hash_key;
+                capture_str_data(key.to_string());
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_CHECK_AND_MUTATE) {
+                dsn::apps::check_and_mutate_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.hash_key;
+                capture_str_data(key.to_string());
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_PUT) {
+                dsn::apps::update_request thrift_request;
+                unmarshall(requests[i], thrift_request);
+                requests[i]->restore_read();
+                key = thrift_request.key;
+                capture_blob_data(key);
+            }
+        }
+    }
+}
+
+void hotkey_collector::capture_blob_data(const ::dsn::blob &key)
+{
+    if (_collector_state.load() == STOP) {
+        return;
+    }
+    std::string hash_key, sort_key;
+    if (key.length() < 2)
+        return;
+    pegasus_restore_key(key, hash_key, sort_key);
+    capture_str_data(hash_key);
+}
+
+void hotkey_collector::capture_str_data(const std::string &data)
+{
+    if (_collector_state.load() == STOP || data.length() == 0) {
+        return;
+    }
+    if (_collector_state.load() == COARSE && _coarse_data_collector != nullptr) {
+        _coarse_data_collector->capture_coarse_data(data);
+    }
+    if (_collector_state.load() == FINE && _fine_data_collector != nullptr) {
+        _fine_data_collector->capture_fine_data(data);
+    }
+    if (_collector_state.load() == FINISH) {
+        return;
+    }
+}
+
+void hotkey_collector::analyse_data()
+{
+    if (_collector_state.load() == STOP) {
+        return;
+    }
```

Comment:
Remove

Reply:
These codes shouldn't be removed.
When `STOP`, `analyse_data()` should return directly, or it would make some errors.